BB-377: use LifecycleTaskV2 for eligible buckets

alexanderchan-scality opened the pull request:
* add helper method for creating LifecycleTasks
** enable V2 listing for bucket containing V2 indexes

bert-e commented:
# Hello alexanderchan-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-377 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.12`


Please check the `Fix Version/s` of BB-377, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-377 contains:


* `8.6.11`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.12`


Please check the `Fix Version/s` of BB-377, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-377 contains:


* `8.6.12`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.13`


Please check the `Fix Version/s` of BB-377, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

rachedbenmustapha commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -304,28 +305,92 @@ class LifecycleBucketProcessor {
+                this._log,
+                (err, bucketIndexes) => {
+                    if (err) {
+                        this._log.warn('unable tparamso retrieve indexing info; using LifecycleTaskV1', {
```
```suggestion
                        this._log.warn('unable to retrieve indexing info; using LifecycleTaskV1', {
```

rachedbenmustapha commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -304,28 +305,92 @@ class LifecycleBucketProcessor {
+                        return next(null);
+                    }
+
+                    isV2 = indexesForFeature.lifecycle.v2.every(
```
Does this ensure that the index is fully "built" (making listing efficient) or is this true if indexes are still building (which can make listings + building very expensive).

alexanderchan-scality replied:
the `getIndexes` method only return indexes that have been built

rachedbenmustapha commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -304,28 +305,92 @@ class LifecycleBucketProcessor {
+
+    }
+
+    _createLifecycleTask(backbeatMetadataProxy, params, cb) {
```
I don't know how expensive it is to list indexes before each listing call. Is there a way to do this check only once before the first listing of the chain?

alexanderchan-scality replied:
indexing should not be expensive to query since most of it would be in memory, but there could be some reads from disk; this could also done as part of the conductor instead, so I  will check if it possible to perform the query there as opposed to checking per task in the bucket processor

rachedbenmustapha commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -304,28 +305,92 @@ class LifecycleBucketProcessor {
+                }),
+            next => {
+                if (!isV2) {
+                    return backbeatMetadataProxy.putBucketIndexes(
```
Is this serialized (through the kafka message key/partition or otherwise)? Not sure what happens if a bunch of these requests are sent at once.

alexanderchan-scality committed:
update backbeat client

alexanderchan-scality committed:
BB-377: add indexes check for lifecycle extension

* add logic to switch between lifecycle listing algo depending on the
  indexes available
** update lifecycle conductor to perform index checks; if v2 indexes are
available, mark bucket task as v2; otherwise, mark it as v1

alexanderchan-scality committed:
bump arsenal

alexanderchan-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`



Please check the status of the associated issue BB-377.

Goodbye alexanderchan-scality.
